promql/engine.go

```diff
@@ -2903,6 +2903,263 @@ func scalarBinop(op parser.ItemType, lhs, rhs float64) float64 {
 	panic(fmt.Errorf("operator %q not allowed for Scalar operations", op))
 }
 
+func handleInfinityBuckets(b histogram.Bucket[float64], le float64) (float64, float64) {
+	var underCount, bucketMidpoint float64
+
+	if math.IsInf(b.Lower, -1) {
+		switch {
+		case le >= b.Upper:
+			// le is greater than or equal to upper bound. Full count applies
+			underCount = b.Count
+			bucketMidpoint = b.Upper
+		case le < 0:
+			// le is negative and less than zero — nothing to keep.
+			underCount = b.Count * 0.5
+			bucketMidpoint = b.Upper
+		default:
+			// Interpolating with treated lower bound as 0 (linear)
+			fraction := le / b.Upper
+			underCount = b.Count * fraction
+			bucketMidpoint = le / 2
+		}
+		return underCount, bucketMidpoint
+	}
+
+	if math.IsInf(b.Upper, 1) {
+		if le <= b.Lower {
+			underCount = 0
+			bucketMidpoint = b.Lower
+		} else {
+			underCount = b.Count * 0.5
+			bucketMidpoint = b.Lower
+		}
+		return underCount, bucketMidpoint
+	}
+	return underCount, bucketMidpoint
+}
+
+// computeSplit calculates the portion of the bucket's count <= le (trim point).
+func computeSplit(b histogram.Bucket[float64], le float64, isPositive, isCustom bool) float64 {
+	if le <= b.Lower {
+		return 0
+	}
+	if le >= b.Upper {
+		return b.Count
+	}
+
+	var fraction float64
+	switch {
+	case isCustom || (b.Lower <= 0 && b.Upper >= 0):
+		fraction = (le - b.Lower) / (b.Upper - b.Lower)
+	default:
+		// Exponential interpolation
+		logLower := math.Log2(math.Abs(b.Lower))
+		logUpper := math.Log2(math.Abs(b.Upper))
+		logV := math.Log2(math.Abs(le))
+
+		if isPositive {
+			fraction = (logV - logLower) / (logUpper - logLower)
+		} else {
+			fraction = 1 - ((logV - logUpper) / (logLower - logUpper))
+		}
+	}
+
+	underCount := b.Count * fraction
+	return underCount
+}
+
+func computeBucketTrim(op parser.ItemType, b histogram.Bucket[float64], rhs float64, isPositive, isCustomBucket bool) (float64, float64) {
+	if math.IsInf(b.Lower, -1) || math.IsInf(b.Upper, 1) {
+		return handleInfinityBuckets(b, rhs)
+	}
+
+	product := math.Abs(b.Lower) * math.Abs(rhs)
+	underCount := computeSplit(b, rhs, isPositive, isCustomBucket)
+	if op == parser.TRIM_UPPER {
+		return underCount, computeMidpoint(b, product, isCustomBucket, isPositive)
```

Comment:
It seems very suspicious that for custom buckets we use the Lower and Upper bound for the mid point, but for exponential buckets we go from the lower bound to the rhs (or rhs to upper bound) for the mid point calculation.

Feels like one of those is wrong.